File: backend/transcend/src/test/java/code/users/logic/UpdateAvatarTest.java
package code.users.logic;

import static code.users.domain.model.UserFixtures.USER_ID_FIXTURE;
import static code.users.domain.model.UserFixtures.aDaoUser;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import code.users.domain.exceptions.UserNotFoundException;
import code.users.domain.model.User;
import code.users.ports.in.UpdateAvatarUseCase;
import code.users.ports.in.UpdateAvatarUseCase.UpdateAvatarCommand;
import code.users.ports.out.UserDao;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;

@SpringJUnitConfig(UpdateAvatarTest.UpdateAvatarTestConfig.class)
@RequiredArgsConstructor(onConstructor_ = {@Autowired})
class UpdateAvatarTest {

  @Configuration
  @Import(UpdateAvatar.class)
  static class UpdateAvatarTestConfig {}

  private final UpdateAvatarUseCase service;

  @MockitoBean private UserDao userDao;

  @Test
  void updatesAvatarSuccessfully() {
    // given
    var user = aDaoUser();
    when(userDao.findById(USER_ID_FIXTURE)).thenReturn(Optional.of(user));

    var command = new UpdateAvatarCommand("test.png", new byte[] {1, 2, 3});

    // when
    service.updateAvatar(USER_ID_FIXTURE, command);

    // then
    ArgumentCaptor<User> captor = ArgumentCaptor.forClass(User.class);
    verify(userDao).updateUser(captor.capture());

    var savedUser = captor.getValue();
    assertThat(savedUser.getDetails().getAvatarId()).isNotNull();
  }

  @Test
  void throwsUserNotFoundException() {
    // given
    var command = new UpdateAvatarCommand("test.png", new byte[] {1, 2, 3});

    when(userDao.findById(USER_ID_FIXTURE)).thenReturn(Optional.empty());

    // when & then
    assertThrows(UserNotFoundException.class, () -> service.updateAvatar(USER_ID_FIXTURE, command));
  }
}

File: backend/transcend/src/test/kotlin/code/users/logic/UpdateAvatarTest.kt
package code.users.logic

import code.users.bootstrap.UserDaoTestSupport
import code.users.domain.model.UserFixtures.NON_EXISTENT_USER
import code.users.domain.model.UserFixtures.USER_ID_FIXTURE
import code.users.ports.`in`.UpdateAvatarUseCase
import code.users.ports.`in`.UpdateAvatarUseCase.UpdateAvatarCommand
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import org.springframework.context.annotation.Import
import org.springframework.security.authorization.AuthorizationDeniedException

@Import(UpdateAvatar::class)
class UpdateAvatarTest(private val service: UpdateAvatarUseCase) : UserDaoTestSupport() {

  init {

    Given("an existing user in the database") {
      val content = byteArrayOf(1, 2, 3)
      val command = UpdateAvatarCommand("test.png", content)

      When("the update avatar service is executed") {
        service.updateAvatar(USER_ID_FIXTURE, command)

        Then("it should successfully update the user's avatar in the database") {
          val savedDetails = userDao.findUserDetailsById(USER_ID_FIXTURE).orElseThrow()
          val newAvatarId = savedDetails.avatarId

          newAvatarId.shouldNotBeNull()

          val savedAvatar = userDao.findById(newAvatarId).orElseThrow()
          savedAvatar.content() shouldBe content
        }
      }
    }

    Given("a non-existent user ID for avatar update") {
      val command = UpdateAvatarCommand("test.png", byteArrayOf(1, 2, 3))

      When("the update avatar service is executed") {
        Then("it should throw a AuthorizationDeniedException") {
          shouldThrow<AuthorizationDeniedException> {
            service.updateAvatar(NON_EXISTENT_USER, command)
          }
        }
      }
    }
  }
}
